sys/arm64/arm64/cpufunc_asm.S

```diff
@@ -50,34 +50,42 @@ __FBSDID("$FreeBSD$");
 	.word	PAGE_MASK
 
 /*
- * Macro to handle the cache. This takes the start address in x0, length
- * in x1. It will corrupt x0, x1, x2, x3, and x4.
+ * Macro to handle the cache. This takes the start address in x/c0, length
+ * in x1. It will corrupt x0, x1, x2, x3, x4 and x6.
  */
 .macro cache_handle_range dcop = 0, ic = 0, icop = 0
+#ifdef __CHERI_PURE_CAPABILITY__
+	mov	c6, c0
+	gcvalue	x0, c6
+#endif
 .if \ic == 0
-	ldr	x3, =dcache_line_size	/* Load the D cache line size */
+	/* Load the D cache line size */
+	LDR_LABEL(x3, PTR(4), dcache_line_size)
 .else
-	ldr	x3, =idcache_line_size	/* Load the I & D cache line size */
+	/* Load the I & D cache line size */
+	LDR_LABEL(x3, PTR(4), idcache_line_size)
 .endif
-	ldr	x3, [x3]
 	sub	x4, x3, #1		/* Get the address mask */
 	and	x2, x0, x4		/* Get the low bits of the address */
 	add	x1, x1, x2		/* Add these to the size */
 	bic	x0, x0, x4		/* Clear the low bit of the address */
```

Comment:
This may end up biting us, as it takes cache line unaligned capabilities out of bounds.

Reply:
The alternative would be to invalidate by VA instead of capability by switching out C64, however I think that to invalidate the cache you should have a valid capability for the VA range. We can add an assertion for this. Even in MIPS where we run into this issue for `bus_dmamap_sync()` we have to rebuild a valid capability anyway to preserve the data in the cache line outside of the capability bounds IIRC.

Reply:
Not if you have a NULL DDC (at least I hope the architecture enforces that...)

Reply:
Right, so we would need to re-derive anyway